spring-resttemplate-2/src/test/java/com/baeldung/resttemplate/json/consumer/service/UserConsumerServiceImplUnitTest.java

```diff
@@ -0,0 +1,148 @@
+                                    .body(USER_JSON)
+                );
+        List<String> actual = tested.processUserDataFromUserArray();
+
+        // Then
+        mockServer.verify();
+        assertEquals(actual.size(), expected.size());
+        assertThat(actual).containsExactly(expected.get(0), expected.get(1));
+    }
+
+    @Test
+    public void whenProcessUserDataAsList_thenOK() {
+        // Given
+        String url = "http://localhost:8080/users";
+        List<String> expected = Arrays.asList("user1", "user2");
+
+        // When
+        mockServer.expect(ExpectedCount.once(),
+                requestTo(url))
+                .andExpect(method(HttpMethod.GET))
+                .andRespond(withStatus(HttpStatus.OK)
+                                    .contentType(MediaType.APPLICATION_JSON)
+                                    .body(USER_JSON)
+                );
+        List<String> actual = tested.processUserDataFromUserList();
+
+        // Then
+        mockServer.verify();
+        assertEquals(actual.size(), expected.size());
+        assertThat(actual).containsExactly(expected.get(0), expected.get(1));
+    }
+
+
+    @Test
+    public void whenProcessNestedUserDataFromArray_thenOK() {
+        // Given
+        String url = "http://localhost:8080/users";
+        List<String> expected = Arrays.asList("user1_address1_postCode", "user1_address2_postCode", "user2_address1_postCode");
+
+        // When
+        mockServer.expect(ExpectedCount.once(),
+                requestTo(url))
+                .andExpect(method(HttpMethod.GET))
+                .andRespond(withStatus(HttpStatus.OK)
+                                    .contentType(MediaType.APPLICATION_JSON)
+                                    .body(USER_JSON)
+                );
+        List<String> actual = tested.processNestedUserDataFromUserArray();
+
+        // Then
+        mockServer.verify();
+        assertEquals(actual.size(), expected.size());
+        assertThat(actual).containsExactly(expected.get(0), expected.get(1), expected.get(2));
+    }
+
+    @Test
+    public void whenProcessNestedUserDataFromList_thenOK() {
+        // Given
+        String url = "http://localhost:8080/users";
+        List<String> expected = Arrays.asList("user1_address1_postCode", "user1_address2_postCode", "user2_address1_postCode");
+
+        // When
+        mockServer.expect(ExpectedCount.once(),
+                requestTo(url))
+                .andExpect(method(HttpMethod.GET))
+                .andRespond(withStatus(HttpStatus.OK)
+                                    .contentType(MediaType.APPLICATION_JSON)
+                                    .body(USER_JSON)
+                );
+        List<String> actual = tested.processNestedUserDataFromUserList();
+
+        // Then
+        mockServer.verify();
+        assertEquals(actual.size(), expected.size());
+        assertThat(actual).containsExactly(expected.get(0), expected.get(1), expected.get(2));
```

Comment:
A couple of mild stylistic points here:

1. `// Given` and `// When` and `// Then` appeared somewhere on the internet as a commenting standard in tests, and on the whole it adds very little. It may be good to finish the sentence `// Given a user service that returns three users` but otherwise, it's just boilerplate

_Small aside here, your `mockServer.expect` is actually part of your `Given` clause in the above, but appears as a `When`!_

2. There's no need to assert both the size and the exact content of a list - the second automatically implies the first.
2.5. If you're using `assertJ`, and I recommend it a lot, then let's not mix it with JUnit's own `assertEquals`